File: 241029/그래프 탐색/graph-traversal.cpp
#include <iostream>
#define MAX_N 1000
using namespace std;

int graph[MAX_N + 1][MAX_N + 1];
bool visited[MAX_N + 1];
int N, M;
int ans;

void dfs(int vertex) {
    for (int i = 1; i <= N; i++) {
        if (graph[vertex][i] && !visited[i]) {
            visited[i] = true;
            ans++;
            dfs(i);
        }
    }
}

int main() {
    int x, y;
    cin >> N >> M;
    
    for (int i = 0; i < M; i++) {
        cin >> x >> y;
        graph[x][y] = 1;
        graph[y][x] = 1;
    }
    
    visited[1] = true;
    dfs(1);
    
    cout << ans;
    
    return 0;
}

File: 250327/그래프 탐색/graph-traversal.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const edges = [];
for (let i = 1; i <= m; i++) {
    edges.push(input[i].split(' ').map(Number));
}


const graph = Array.from(Array(1001), () => Array());
const visited = Array(1001).fill(false);

edges.forEach(([x, y]) => {
    graph[x].push(y);
    graph[y].push(x);
});

let ans = 0;
visited[1] = true;

function dfs(vertex) {
    graph[vertex].forEach((curV) => {
        if (!visited[curV]) {
            ans++;
            visited[curV] = true;
            dfs(curV);
        }
    });
}

dfs(1);
console.log(ans);